rpc/src/server/filter.rs

```diff
@@ -34,34 +35,49 @@ where
     channels_per_key: u32,
     dropped_keys: mpsc::UnboundedReceiver<K>,
     dropped_keys_tx: mpsc::UnboundedSender<K>,
-    key_counts: FnvHashMap<K, Weak<Tracker<K>>>,
+    key_counts: FnvHashMap<K, TrackerPrototype<K>>,
     keymaker: F,
 }
 
 /// A channel that is tracked by a ChannelFilter.
 #[derive(Debug)]
 pub struct TrackedChannel<C, K> {
     inner: C,
-    tracker: Arc<Tracker<K>>,
+    tracker: Tracker<K>,
 }
 
 impl<C, K> TrackedChannel<C, K> {
     unsafe_pinned!(inner: C);
 }
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 struct Tracker<K> {
-    key: Option<K>,
+    key: Option<Arc<K>>,
+    counter: Counter,
     dropped_keys: mpsc::UnboundedSender<K>,
 }
 
 impl<K> Drop for Tracker<K> {
     fn drop(&mut self) {
-        // Don't care if the listener is dropped.
-        let _ = self.dropped_keys.unbounded_send(self.key.take().unwrap());
+        if self.counter.count() <= 1 {
+            // Don't care if the listener is dropped.
+            match Arc::try_unwrap(self.key.take().unwrap()) {
+                Ok(key) => {
+                    let _ = self.dropped_keys.unbounded_send(key);
+                }
+                _ => unreachable!(),
```

Comment:
Not: instead of the match, how about calling `Arc::try_unwrap(...).expect("nice error message")`?

Reply:
That would require K to be Debug